base/strings/styled.jl

```diff
@@ -0,0 +1,365 @@
+
+# Constructors to avoid recursive wrapping
+
+StyledString(s::StyledString, props::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}) =
+    StyledString(s.string, vcat(s.properties, props))
+
+StyledChar(c::StyledChar, props::Vector{Pair{Symbol, Any}}) =
+    StyledChar(c.char, vcat(s.properties, props))
+
+# To avoid pointless overhead
+String(s::StyledString{String}) = s.string
+
+## Conversion/promotion ##
+
+convert(::Type{StyledString}, s::StyledString) = s
+convert(::Type{StyledString{S}}, s::S) where {S <: AbstractString} =
+    StyledString(s, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+convert(::Type{StyledString}, s::S) where {S <: AbstractString} =
+    convert(StyledString{S}, s)
+StyledString(s::S) where {S <: AbstractString} = convert(StyledString{S}, s)
+
+convert(::Type{StyledChar}, c::StyledChar) = c
+convert(::Type{StyledChar{C}}, c::C) where { C <: AbstractChar } =
+    StyledChar{C}(c, Vector{Pair{Symbol, Any}}())
+convert(::Type{StyledChar}, c::C) where { C <: AbstractChar } =
+    convert(StyledChar{C}, c)
+
+StyledChar(c::AbstractChar) = convert(StyledChar, c)
+StyledChar(c::UInt32) = convert(StyledChar, Char(c))
+StyledChar{C}(c::UInt32) where {C <: AbstractChar} = convert(StyledChar, C(c))
+
+promote_rule(::Type{<:StyledString}, ::Type{<:AbstractString}) = StyledString
+
+## AbstractString interface ##
+
+ncodeunits(s::StyledString) = ncodeunits(s.string)
+codeunits(s::StyledString) = codeunits(s.string)
+codeunit(s::StyledString) = codeunit(s.string)
+codeunit(s::StyledString, i::Integer) = codeunit(s.string, i)
+isvalid(s::StyledString, i::Integer) = isvalid(s.string, i)
+@propagate_inbounds iterate(s::StyledString, i::Integer=firstindex(s)) =
+    if i <= lastindex(s.string); (s[i], nextind(s, i)) end
+eltype(::Type{<:StyledString{S}}) where {S} = StyledChar{eltype(S)}
+firstindex(s::StyledString) = firstindex(s.string)
+lastindex(s::StyledString) = lastindex(s.string)
+
+function getindex(s::StyledString, i::Integer)
+    @boundscheck checkbounds(s, i)
+    @inbounds if isvalid(s, i)
+        StyledChar(s.string[i], textproperties(s, i))
+    else
+        string_index_err(s, i)
+    end
+end
+
+## AbstractChar interface ##
+
+ncodeunits(c::StyledChar) = ncodeunits(c.char)
+codepoint(c::StyledChar) = codepoint(c.char)
+
+# Avoid the iteration fallback with comparison
+cmp(a::StyledString, b::AbstractString) = cmp(a.string, b)
+cmp(a::AbstractString, b::StyledString) = cmp(a, b.string)
+
+## Concatenation ##
+
+function *(a::StyledString, b::StyledString)
+    combinedproperties = copy(a.properties)
+    append!(combinedproperties,
+            [(start + lastindex(a) => stop + lastindex(a), prop)
+             for ((start, stop), prop) in b.properties])
+    StyledString(a.string * b.string, combinedproperties)
+end
+
+*(a::StyledString, b::AbstractString) = a * convert(StyledString, b)
+*(a::AbstractString, b::StyledString) = convert(StyledString, a) * b
```

Comment:
Rather than having these two, it feels like a `promote_type` method might be better, although I suppose this avoids allocating a new `StyledString` object every time you want to concatenate.

Reply:
I don't think this comment should affect merging this or not; can be adjusted later.

Reply:
This is obsolete with the adjustment to the `string` method made.